Snowiiii/Pumpkin
# Correct Interactive Block Behavior

## Description
Fixes #275 and many other interactive block issues:
- All interactive blocks are per block now (for instance, chests and furnaces do not share containers globally)
- Crafting table containers are both unique per player AND per block
- Breaking interactive blocks removes all contents (can drop items once item drop is implemented)
- All active container users are forcibly closed when another `player` removes a block
- Removed artificial cap on container screen limit which kicked players after opening a container 15 times (we may not be handling container ids correctly)
- Added a `standard` set of container functions to reduce code duplication for interactive blocks

This PR requires #384 


## Testing
Two clients
- Place a crafting table, a furnace, and a chest
- Open each and place items in container
- Open crafting table again, notice no items are stored
- Open all other blocks and see items in both clients
- Open crafting table in both and notice containers are not shared

Internally, this change maintains the OpenContainer HashMap and a new AtomicU32 which tracks the current index for the generation of uuids. On large persistent servers, unique containers could be optimized to reduce memory usage.  Every user who opens a unique container will generate and cache a new data structure (think 10,000 users opening a crafting table, enderchest, etc). Future work could include separating unique containers (crafting table) and persistent unique containers (ender chest) into separate data structures (ex. stacks instead of maps).

Also, another optimization could be made to client interactive block interactions. Currently, this requires that the server iterate through the entirety of the OpenContainer HashMap. It may be better to create a sort of registry which maps block position to OpenContainerID in the long term (again large server optimizations). Otherwise, the algorithm is randomized brute force with a time complexity of O(n + m).


## Checklist
Things need to be done before this Pull Request can be merged.

- [x] All unit tests pass: `cargo test`
- [ ] I added new unit tests, so other people don't accidentally break my code by changing other parts of the codebase. [How?](https://doc.rust-lang.org/book/ch11-01-writing-tests.html)
